packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -564,298 +705,303 @@ function condenseErrorsAndFilterSiblings<T>(
 		// and look sth like this:
 		// Missing either keys ("A", "B" and "C"), ("A", and "D"), or "F"
 		errors.push(
-			...validDefinitions.map(d => ({
-				kind: 'some_missing_keys' as 'some_missing_keys',
-				node: d.node.runtimeNode.node,
-			})),
+			...validDefinitions
+				.flatMap(d =>
+					d.errors
+						.filter(e =>
+							e.kind === 'missing_key' &&
+							!alwaysMissing.some(oe =>
+								oe.node.originalNode === e.node.originalNode
+							)
+						)
+				)
+				.map(e => e.node as RuntimeNode<T>)
+				.concat(
+					validDefinitions.flatMap(d => d.errors).filter(e =>
+						e.kind === 'some_missing_keys'
+					).map(e => e.node as RuntimeNode<T>),
+				)
+				.filter((v, i, arr) =>
+					arr.findIndex(o => o.originalNode === v.originalNode) === i
+				)
+				.map(n => ({
+					kind: 'some_missing_keys' as 'some_missing_keys',
+					node: n,
+				})),
 		)
 	}
 
 	// TODO handle list length range and value range errors (merge ranges, could be multiple possible distinct ranges)
 
 	return {
-		siblings: validDefinitions.map(d => d.node),
+		definitions: validDefinitions.map(d => d.definition),
 		condensedErrors: errors,
 	}
 }
 
-function check<T>(
-	node: CheckerTreeEntryNode<T>,
+interface ValidDefintionResult<T> {
+	errors: McdocCheckerError<T>[]
+	childDefinitions: (McdocType | undefined)[]
+}
+
+function checkShallowly<T>(
+	runtimeNode: RuntimeNode<T>,
+	simplifiedInferred: SimplifiedMcdocTypeNoUnion,
+	children: RuntimeUnion<T>[],
+	typeDef: SimplifiedMcdocTypeNoUnion,
 	options: McdocCheckerOptions<T>,
```

Comment:
I very much like that you made this function not use any of the graph types, and also got rid of the mutations/side effects!